File: app/src/main/java/com/icostel/arhitecturesample/di/components/AppComponent.java
package com.icostel.arhitecturesample.di.components;

import android.app.Application;

import com.icostel.arhitecturesample.SampleApp;
import com.icostel.arhitecturesample.di.modules.ActivityModule;
import com.icostel.arhitecturesample.di.modules.AppModule;

import javax.inject.Singleton;

import dagger.BindsInstance;
import dagger.Component;
import dagger.android.support.AndroidSupportInjectionModule;

@Singleton
@Component(
        modules = {
                AndroidSupportInjectionModule.class,
                AppModule.class,
                ActivityModule.class
        }
)

public interface AppComponent {

    @Component.Builder
    interface Builder {
        @BindsInstance
        Builder application(Application application);

        AppComponent build();
    }

    void inject(SampleApp sampleApp);
}

File: app/src/main/java/com/icostel/arhitecturesample/di/components/AppComponent.kt
package com.icostel.arhitecturesample.di.components

import com.icostel.arhitecturesample.SampleApp
import com.icostel.arhitecturesample.di.modules.ActivityModule
import com.icostel.arhitecturesample.di.modules.AppModule

import javax.inject.Singleton

import dagger.Component
import dagger.android.AndroidInjectionModule
import dagger.android.AndroidInjector

@Singleton
@Component(modules = [
    AndroidInjectionModule::class,
    AppModule::class,
    ActivityModule::class
])
interface AppComponent: AndroidInjector<SampleApp> {

    @Component.Factory
    abstract class Factory : AndroidInjector.Factory<SampleApp>
}
